lib/kino.ex

```diff
@@ -167,37 +167,93 @@ defmodule Kino do
-  the accumulated value. The callback should return either of:
-
-    * `{:cont, term_to_render, acc}` - the continue
-
-    * `:halt` - to no longer schedule callback evaluation
+  The callback receives a stream element and should return a term
+  to be rendered.
 
   This function uses `Kino.Frame` as the underlying kino.
   It returns nothing (a non-printable result).
 
   ## Examples
 
+  An animation is created by turning a stream of values into
+  subsequent animation frames:
+
+      Stream.interval(100)
+      |> Stream.take(100)
+      |> Kino.animate(fn i ->
+        Kino.Markdown.new("**Iteration: `#{i}`**")
+      end)
+
+  Alternatively an integer may be passed as a shorthand for
+  `Stream.interval/1`:
+
       # Render new Markdown every 100ms
-      Kino.animate(100, 0, fn i ->
-        md = Kino.Markdown.new("**Iteration: `#{i}`**")
-        {:cont, md, i + 1}
+      Kino.animate(100, fn i ->
+        Kino.Markdown.new("**Iteration: `#{i}`**")
+      end)
+  """
+  @spec animate(Enumerable.t() | pos_integer(), (term() -> any())) :: nothing()
+  def animate(stream_or_interval_ms, fun) when is_function(fun, 1) do
+    animate(stream_or_interval_ms, nil, fn item, nil ->
+      {:cont, fun.(item), nil}
+    end)
+  end
+
+  @doc ~S"""
+  A stateful version of `animate/2`.
+
+  The callback receives a stream element and the accumulated state
+  and it should return either of:
+
+    * `{:cont, term_to_render, state}` - the continue
+
+    * `:halt` - to no longer schedule callback evaluation
+
+  ## Examples
+
+  This function is primarily useful to consume `Kino.Control` events:
+
+      button = Kino.Control.button("Click")
+
+      button
+      |> Kino.Control.stream()
+      |> Kino.animate(0, fn _event, counter ->
+        new_counter = counter + 1
+        md = Kino.Markdown.new("**Clicks: `#{new_counter}`**")
+        {:cont, md, new_counter}
       end)
   """
   @spec animate(
-          pos_integer(),
-          term(),
-          (term() -> {:cont, term(), acc :: term()} | :halt)
+          Enumerable.t() | pos_integer(),
+          state,
+          (term(), state -> {:cont, term(), state} | :halt)
         ) :: nothing()
-  def animate(interval_ms, acc, fun) do
+        when state: term()
+  def animate(stream_or_interval_ms, state, fun)
+
+  def animate(interval_ms, state, fun) when is_integer(interval_ms) do
     frame = Kino.Frame.new()
 
-    Kino.Frame.periodically(frame, interval_ms, acc, fn acc ->
-      case fun.(acc) do
-        {:cont, term, acc} ->
+    fun =
+      cond do
+        is_function(fun, 1) ->
+          # TODO: remove on Kino v0.8
+          IO.warn(
+            "Passing arity-1 function to Kino.animate/3 is deprecated, " <>
+              "please use Kino.animate/2 or pass an arity-2 function"
+          )
+
+          fn _i, state -> fun.(state) end
+
+        is_function(fun, 2) ->
+          fun
+      end
+
+    Kino.Frame.periodically(frame, interval_ms, {0, state}, fn {i, state} ->
```

Comment:
We could just call `animate(Stream.interval(interval_ms), state, fun)` instead, but in this specific case we can re-use the frame process, so it's probably worth it.